### ISSUE TITLE
Proposal to add Checkpointing to streamz.

### ISSUE THREAD

**Issue by chinmaychandak** [python-streamz/streamz#278]
I closed out the earlier checkpointing PR #272, since it was an incorrect implementation. Before I raise a new PR, I wanted to put out a proposal and get as many reviews as possible, so as to have a sure footing to implement the feature. So here goes:

The **motivation** for Checkpointing is simple — we need to track the amount of data that has been **processed** in our streaming pipeline so that in case the stream has to restart, it can pick up from where it left off. Please note, that **processed** here is different from data only being read/consumed in the pipeline. 

**Spark follows the following heuristics for checkpointing:**
1. Checkpointing is mandatory for all streaming pipelines, as would be in any production environments. _However_, we can make checkpointing optional in streamz API, if need be.
2. Whenever a batch of data is processed completely, and the results are written to a `sink`, Spark creates a new checkpoint for this batch with the latest offsets for all partitions for the topic(s) involved in the batch. 
3. Spark does not consider printing to console or the like as a `sink`. In case of streamz, the only valid sink for now is Kafka.
4. The directory structure of checkpointing is as follows:
```
checkpoint_path/ #this is provided by the user 
      topic1/
                    1.txt #Checkpoint for the first batch
                    2.txt #Checkpoint for the second batch 
                    .
                    .
      topic2/
                    1.txt
                    .
```
5. Now, if a stream is killed for some reason, and restarts, it would pick up reading the Kafka topic from the last checkpoint file created. 
6. Only the last few checkpoints are maintained, and the older ones are deleted. We can hard-code this number to something, for example, 100.
7. If multiple streams join/union, Spark also takes care of the checkpoints for all topics that may be involved.

**Implementation in streamz:**
The following is one of the cleanest ways, IMO, to implement checkpointing.

1. Reading a batch from Kafka starts off with `from_kafka_batched()`.
```
from_kafka_batched([topics], consumer_conf, checkpoint_path, ..):	
	consumer_conf[‘enable.auto.commit’] = False
	for topic in topics:
		Retrieve last created checkpoint for the topic from the checkpoint_path.
 		for partition in partitions:
			emit (partition, low, high) tuples in stream for get_message_batch()
``` 
Note that, `[topics]` parameters is a list of topic that the stream can now subscribe/consume to/from. 
The `checkpoint_path` parameter has to be specified by the user. In case the user does not want to use checkpointing, we can set the default to `None` to use streamz as is.

2. `get_message_batch()` consumes the messages from the specific partition between the low and high offsets specified in the tuples emitted in the stream.
```
get_message_batch(kafka_params, topic, partition, low, high, checkpoint_path, ..):
	Consume messages
        Create a dirty-bit file in the checkpoint_path — this is an indication that some data is yet to be checkpointed.
	Create/update a write-ahead log (WAL) textfile for this topic partition with the latest offsets that have been read and are ready to be `checkpointed` if processing is successful.      
```
The purpose of the dirty-bit is explained later on.
The WAL would be a text-file, `#partition.txt` in a topic-level directory in the checkpoint path.
It would contain the offset for this topic partition that has last been consumed, and if processing goes through to finally write back to a sink, this the offset to be checkpointed.

The directory structure would now look like this:
```
checkpoint_path/
      dirty_bit
      write-ahead-logs/
           topic1/
                    1.txt #Log for first partition of topic 1.
                    2.txt #Log for second partition of topic 1. 
                    .
                    .
           topic2/
                    1.txt
                    .
      topic1/
                    1.txt #Checkpoint for the first batch
                    2.txt #Checkpoint for the second batch 
                    .
                    .
      topic2/
                    1.txt
                    .
```

3. There would now be some stream-specific `processing` functions, e.g., `map()`, etc.
4.  Finally, if we write back the results to Kafka using `to_kafka()`, this is where the checkpoint creation would take place.
```
to_kafka(checkpoint_path, output_topic, producer_conf):
       ... #The current operations within to_kafka() stay as is.
       
       #Everything below will be done from a separate function add_checkpoint()
       If no dirty-bit file, exit.
       Delete dirty-bit file here.
        new_checkpoint = {}
	for topic in [topics from WAL]: (#We can make this multi-threaded)
		for partition in WAL/topic path:
                      new_checkpoint[partition] = offset
                Create checkpoint file for this batch for this topic at `checkpoint_path/topic`. 
```
A checkpoint file would look like a JSON {partition:offset} dictionary:
```
checkpoint_path/topic1/1.txt
{0: 5, 1: 4, 2: 3}
```
Let's say we decide to keep only the last 100 checkpoints for each topic. Then, as soon as the 101st checkpoint is created, the 1st checkpoint 1.txt is deleted, and the 101.txt gets created. 

This finishes the implementation. 

Now, a few questions and their answers:
1. Since, `to_kafka()` is called on every element of the stream, and if we are using Dask, all WAL for all partitions are created in parallel. After we have `gather().to_kafka()`, the checkpoint is created on the first call to `to_kafka()` itself. In this case, how do we avoid duplicate checkpoints being created? A maximum of #partitions duplicate checkpoints can be created.
Answer: This is where the dirty bit file comes into play. If new data is consumed, a dirty bit is set. `to_kafka()` does not attempt to create a new checkpoint if the dirty bit is not set. After a checkpoint is created, the dirty bit is deleted by `to_kafka()`.

2. The WAL and the checkpoints would be stored on the local file system for now?
Answer: Yes, we can support S3 in the future, like Spark.

3. In the future, if we have different sinks apart from Kafka, like SQS, Google Pub-Sub, ElasticSearch, do we need to implement checkpointing all over again?
Answer: No. When writing to a sink, one would just call the `add_checkpoint()` function that we would create as part of this PR. **The source would always be Kafka, though, for the foreseeable future.**  

4. We need to pass the `checkpoint_path` to both `from_kafka_batched()` and `to_kafka()`? 
Answer: Yes, that's unfortunate. I'd love to hear another way to do this for a cleaner API. But we need the path in `from_kafka_batched()` to retrieve the last checkpoint, and in `to_kafka()` to create a new checkpoint. 

5. Would this work with Dask?
Answer: Absolutely. We would definitely want this to work with Dask!

**Comment by chinmaychandak** [comment]
@martindurant, @CJ-Wright, Could you please review this?

**Comment by jsmaupin** [comment]
There are two things I think could exist in Streamz that would simplify this greatly.
1. Message metadata
1. Stream context

When a new Stream object is created or another Stream object is attached to the end of a stream, it can add its relevant parts to the context and the context would get passed downstream as each now Stream object is attached to the stream.  `from_kafka_batched` would attach a callback to the context that takes in the message metadata and knows how to write kafka offsets to some store.

**Comment by jsmaupin** [comment]
Also, if we had an idea of an OutputStream that would have the functionality to write this metadata to the callback that would be helpful.  It would sit on the inheritance tree between Stream and all "sink-like" streams, such as `to_kafka`.

**Comment by jsmaupin** [comment]
There are two things I think could exist in Streamz that would simplify this greatly.

1. Message metadata
1. Stream context

When a new Stream object is created or another Stream object is attached to the end of a stream, it can add its relevant parts to the context and the context would get passed downstream as each now Stream object is attached to the stream.  `from_kafka_batched` would attach a callback to the context that takes in the message metadata and knows how to write kafka offsets to some store.

**Comment by jsmaupin** [comment]
Also, if we had an idea of an OutputStream that would have the functionality to write this metadata to the callback that would be helpful. It would sit on the inheritance tree between Stream and all “sink-like” streams, such as to_kafka.

**Comment by chinmaychandak** [comment]
@martindurant, @CJ-Wright Did you guys have a chance to look at this?

**Comment by CJ-Wright** [comment]
I did get a chance to look, although I don't know if I have thought enough about it to have a coherent opinion yet. One of my concerns is that I'd want sources to be black boxes which provide data, the specific implementation of things is not something we should expose to the user. Would it be possible for this work in a backend agnostic way?

**Comment by CJ-Wright** [comment]
> When a new Stream object is created or another Stream object is attached to the end of a stream, it can add its relevant parts to the context and the context would get passed downstream as each now Stream object is attached to the stream. from_kafka_batched would attach a callback to the context that takes in the message metadata and knows how to write kafka offsets to some store.

I think things like this would be best set up as graph walkers. In this way you create a streamz graph, and then have a walker come in and modify (decorate?) the way the stream nodes operate.

**Comment by chinmaychandak** [comment]
> sources to be black boxes
The only source for checkpointing here, is Kafka. By black boxes, do you mean we should not have the checkpointing path accepted as part of the API?

> specific implementation of things is not something we should expose to the user
Nothing related to checkpointing would be exposed as an API to users; everything happens behind the scene. The only thing is that the user would pass a checkpointing path in `from_kafka_batched()`.

**Comment by CJ-Wright** [comment]
By black boxes I mean that the source API, esp. what methods they expose, should ideally be the same across all sources. That way one can exchange Kafka for ZMQ with no changes to the source code, aside from changing the name of the source class.

**Comment by jsmaupin** [comment]
I have to agree with the idea of making these black box abstractions.  Each source could have their own unique way of writing checkpoints. The downstream doesn't have to know how the checkpoints are written, they only have to know the name of the method that makes it happen.

Just so I understand @CJ-Wright, what you're saying is that we have a walkable DAG where we can inject mix-ins at runtime?

**Comment by CJ-Wright** [comment]
Walking the DAG is a potential approach, we discuss it some in https://github.com/python-streamz/streamz/issues/264. I think the main bit of code is here https://github.com/xpdAcq/rapidz/blob/master/rapidz/graph.py#L57 (note that this is in rapidz, not streamz so it would need to be PRed over) which could be modified to be more user friendly (default the graph arg to a empty graph and return it, if needed). This would result in a `networkx` graph of the nodes, which could be walked, inspected, and modified as needed.

This method would also do a nice job of live viz, logging, etc.

**Comment by jsmaupin** [comment]
Hey @CJ-Wright can you help me understand one thing?  Why do we need walkable graph or even something to manage a DAG when in Streamz, everything is pretty in two lists inside the Stream object with the names of `upstreams` and `downstreams`? Can we not just iterate over one of these lists or is it more complicated than that?

**Comment by CJ-Wright** [comment]
You can do that, although everything is not in a single stream's upstreams/downstreams so you will need to run recursively. You most likely want something like: https://github.com/xpdAcq/rapidz/blob/master/rapidz/graph.py#L26 Although at that point you are most of the way to a graph. The graph structure could make some information more available, but if it isn't needed for this application then you don't need to use it.

**Comment by kivo360** [comment]
Any way I can help make this a reality?

**Comment by CJ-Wright** [comment]
@martindurant what do you think?

**Comment by jsmaupin** [comment]
> Any way I can help make this a reality?

For us, the best thing right now is just getting feedback and figuring out if this method is good or not.  And, if not, what is a better strategy.

**Comment by CJ-Wright** [comment]
On a related note would it be possible to use backpressure to your advantage? One might be able to get the results of the `emit` call in the `source` and hand that to the checkpointing system. Eg. if using dask futures you could store the futures and then figure out if the future has completed yet. If the future is done then you can know that that piece of data passed the checkpoint. If you are not using futures then backpressure should be respected and when `emit` returns control of the execution then that piece of data is finished being processed.

**Comment by CJ-Wright** [comment]
One thing that might be tricky is that a future may have finished but not moved past the gather because a longer runnign task is ahead of it in the gather queue. In that case you can check the gather queue and see if the future is in the queue at which point you know that the data is not finished processing yet.

**Comment by jsmaupin** [comment]
Yes, I do think the Dask case is simpler because Dask builds an abstract DAG behind the scenes.  If the Dask worker handles the data from the input to the output, then it becomes simpler to just commit the offsets once the future is marked as done.

However, in the non-Dask case, Streamz creates a concrete pipeline. It becomes a little more tricky unless we modify Streamz to work like Dask in that, it would not statically tie each node in the pipeline to each other, but it would return a future and build the DAG behind the scene somehow.

**Comment by CJ-Wright** [comment]
What does "statically tie" mean?

**Comment by jsmaupin** [comment]
What I mean is, if you use Streamz without Dask, each element of data you place in the beginning of the pipeline is passed directly to each node in the pipeline unless it gets filtered out somehow.  However, with Dask, all the client (or the Streamz library) sees are a set of futures. The actual DAG is abstracted away.

**Comment by martindurant** [comment]
Streamz does do this for any async node. Sync nodes just run in blocking fashion (i.e., after you call emit, if you return,you know the node finished processing).

**Comment by CJ-Wright** [comment]
> Streamz does do this for any async node. Sync nodes just run in blocking fashion (i.e., after you call emit, if you return,you know the node finished processing

I think that if you return then you know the *entire processing for that data element* is finished (if you are watching the source node's `emit` call).

**Comment by CJ-Wright** [comment]
Note that this works even when you have data processing after a `buffer/gather` node that breaks backpressure. Once a future comes of the gather the pipeline is back in blocking mode, you can't insert a new data element into the pipeline until the current data element has finished processing.

**Comment by martindurant** [comment]
> entire processing for that data element is finished

Right - the node blocked, and so did all of its downstreams. If any element is async, then all are.

Also agree on gather - indeed we don't know when the cluster is done with some data, but if we gather and await on that, then we necessarily know the processing finishes.

**Comment by jsmaupin** [comment]
So, are we saying that checkpointing will only work in with the use of Dask or by setting `asynchronous=True`?

**Comment by martindurant** [comment]
My view, kafka should be async or dask, yes - that's independent of checkpointing.

**Comment by CJ-Wright** [comment]
Checkpointing should work for all things, including synchronous streamz. In synchronous streamz you can trigger the checkpoint based on when the `emit` method returns. If it hasn't returned yet then the data hasn't finished yet, when it returns you have finished that data's processing.

**Comment by jsmaupin** [comment]
> Checkpointing should work for all things, including synchronous streamz. In synchronous streamz you can trigger the checkpoint based on when the emit method returns. If it hasn't returned yet then the data hasn't finished yet, when it returns you have finished that data's processing.

Right, Okay! Yes, that makes sense.

**Comment by CJ-Wright** [comment]
To be fair I agree with @martindurant, you most likely want to use some parallel backend to run the computation, unless your kafka stream is much slower than your data processing.

**Comment by CJ-Wright** [comment]
The joining nodes might be a little odd though (`zip`, `combine_latest`). Since you could send data into those nodes and then:
- sync: when the data hits a joining node you may get a return even though that data hasn't "exited" the pipeline, it is effectively sitting in a cache waiting for another piece of data to come down so both can exit together
- async/dask: when the data hits a joining node the future hasn't entered the buffer/gather.

Is this data considered "done" once it hits the cache? Or should we re-run the data through the pipeline before this?

I think this plays more into questions of how do we restock caches in the event of a terminal error?

**Comment by CJ-Wright** [comment]
Not that I think this should be a blocking thing, having checkpointing that works for linear pipelines might be a good start.

To handle the joining node issue, you may need to go back to the "have something inspect the pipeline nodes" approach. Since you inspect the pipeline after it is created you then can know which nodes are join nodes, and then you can keep track of all the caches and determine if a piece of data has left the pipeline and does not need to be any part of a restart.

**Comment by jsmaupin** [comment]
I had assumed that `from_kafka_batched` was async, especially on Dask. After learning more about this, I can see that it is synchronous. This makes checkpointing much simpler here, but it will definitely hurt throughput. I quickly coded a rough solution and it seems to work.

Is there some reason I shouldn't make a pull request using this, maybe with some additional parameters?  I have comments in the parts I added.
```python
    @gen.coroutine
    def poll_kafka(self):
        import confluent_kafka as ck

        # Get the offsets committed in Kafka
        tps = []
        for partition in range(self.npartitions):
            tps.append(ck.TopicPartition(self.topic, partition))

        # Set the current positions to the fetched offsets
        for n, tp in enumerate(self.consumer.committed(tps, timeout=0.1)):
            self.positions[n] = tp.offset

        try:
            while not self.stopped:
                out = []
                for partition in range(self.npartitions):
                    tp = ck.TopicPartition(self.topic, partition, 0)
                    try:
                        low, high = self.consumer.get_watermark_offsets(
                            tp, timeout=0.1)
                    except (RuntimeError, ck.KafkaException):
                        continue
                    current_position = self.positions[partition]
                    lowest = max(current_position, low)
                    if high > lowest:
                        out.append((self.consumer_params, self.topic, partition,
                                    lowest, high - 1))
                        self.positions[partition] = high

                for part in out:
                    yield self._emit(part)
                    
                    # Commit the offsets back to Kafka synchronously for each batch
                    topic, partition, _, offset = part[1:]
                    tp = ck.TopicPartition(topic, partition, offset + 1)
                    self.consumer.commit(offsets=[tp], asynchronous=False)

                else:
                    yield gen.sleep(self.poll_interval)
        finally:
            self.consumer.unsubscribe()
            self.consumer.close()
```

**Comment by CJ-Wright** [comment]
I think you need to handle the case when `yield self._emit(part)` returns a future because the datum was pushed into a Dask enabled pipeline. In this case you need to wait until the future is done and no longer in any `buffer` nodes to know that it has truly exited the pipeline.

**Comment by kivo360** [comment]
Is it possible to use [Dask's callback system](https://docs.dask.org/en/latest/futures.html#distributed.Future.add_done_callback) to give a clue of whether a process is done or not?

My b on the delay. I've been visiting a friend for the last 3 days.

**Comment by martindurant** [comment]
I would discourage starting more threads. Why not await on the future, as with other backpressure?

**Comment by jsmaupin** [comment]
> I think you need to handle the case when yield self._emit(part) returns a future because the datum was pushed into a Dask enabled pipeline.

Isn't the future already awaited by the Tornado library? The `poll_kafka` function is a callback added by the `add_callback` method.  If this was not the case, then the `yield gen.sleep(self.poll_interval)` call a few lines down would not work as intended as it returns a future as well.

The following script does print a line every 1 second. The behavior is the same when the `loop` function yields from `gen.sleep(1)` directly or if it wraps it in a list.

```python
@gen.coroutine
def loop():
    while True:
        print(int(time()))
        yield gen.sleep(1)


async def main():
    IOLoop.current().add_callback(loop)
    while True:
        await gen.sleep(0.1)


if __name__ == '__main__':
    logging.basicConfig(
        format='%(asctime)s %(levelname)s %(name)s - %(message)s',
        level=logging.INFO)
    IOLoop.current().run_sync(main)
```

```
1571676013
1571676014
1571676015
1571676016
1571676017
1571676018
```

**Comment by martindurant** [comment]
I'm not sure what the code above shows, but `emit()` (without the underscore) performs `_emit()` and awaits if the node is running async.

**Comment by jsmaupin** [comment]
Oops, you are correct.  The code above does not show anything.  Would it make sense then to try and await one future for each Dask worker or something similar to the `to_kafka` method where a list is maintained?

**Comment by martindurant** [comment]
I doubt it makes sense to try to figure out which worker is doing what - Dask makes its own decisions about how to parcel out work. However, it does mean that events are not necessarily processed in order, so the checkpointing state needs to keep hold of not only the oldest event still in flight, but of any later events that have already been successfully processed.

**Comment by kivo360** [comment]
I know were adverse to using new threads, though have we looked into dask's [diagnostic tools](https://distributed.dask.org/en/latest/_modules/distributed/diagnostics/plugin.html)?

**Comment by jsmaupin** [comment]
I'm slowly working on completing an implementation that will support checkpointing.  It involves having the source create a future and each component will forward this future to the end of the pipeline.  Depending on how each component functions, it will treat the future differently.  For example, a map() operation will simply pass the future to the downstream. A filter, however, will pass the future downstream only if the data passes the filter.  If the data does not pass the filter, the future will be marked as complete.  

As far as determining where the pipeline ends, If a stream component has no more down-streams, then that is the end of the pipeline, and the future is marked as done.  However, it is possible that a DAG can have multiple outputs. I'm thinking this could be solved by splitting the future into multiple futures where each output will have an option to save a separate set of offsets.  

One worry I previously had about this implementation is the overhead of managing all the futures for each data element.  I believe that the overhead will be limited by several facts.  Firstly, the futures are optional.  Either an empty list or `None` will be passed down the pipeline if no futures are used.  Also, in our case, the data will be handled in frames and batches. This means there will be one future for a large amount of data. 

Please post any questions or thoughts you may have.

**Comment by CJ-Wright** [comment]
FYI currently there is no support for filter with dask futures (I'm hoping there will be some in the future).
I'm confused how `map` doesn't change the future, is this a different future than the dask futures?

How does this handle joining nodes like `combine_latest` and `zip`?

**Comment by martindurant** [comment]
I find it hard to comment on the idea, because I don't really understand. Usually when talking about these streams with back-pressure, we have values going *forward* from one note to another and futures going *backward*; i.e., the upstream receives a future that it awaits on, which is when it can be guaranteed that the passed value has actually been used. I don't get how a source can be creating futures, but I'm prepared to see it!

**Comment by chinmaychandak** [comment]
> I'm confused how `map` doesn't change the future, is this a different future than the dask futures?

The 'futures' created by the source (e.g., futures created in `from_kafka_batched` would be 1 per partition) are different from Dask futures. These are used to track the processing of data elements in the stream pipeline. At the end of the pipeline, once this future gets marked as 'done', streamz can then ensure that it is safe to checkpoint the data since the processing is complete. Thus, in case of `map`, the futures just get passed along.

Please do correct me if I am wrong @jsmaupin.

**Comment by chinmaychandak** [comment]
> the upstream receives a future that it awaits on, which is when it can be guaranteed that the passed value has actually been used

This is exactly what is happening in Jarod's approach!

**Comment by martindurant** [comment]
So it is indeed basically metadata with a callback?

**Comment by chinmaychandak** [comment]
Exactly!
